### PR TITLE
DEV/STYLE: use ruff for linting

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
`flake8` is one of the slow parts of our linting (which can become especially annoying when using it as a pre-commit check), and https://github.com/charliermarsh/ruff/ is a faster replacement, and is starting to picked up by several other projects.

This PR only replaces flake8 and flake8-bugbear on code (not docs) and yesqa with ruff. Later on, we could also consider replacing other parts of our linting with ruff (eg isort, pylint, pyupgrade, or some of our custom checks)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -184,6 +184,20 @@ exclude = '''
 )
 '''
 
+[tool.ruff]
```
Comment (MarcoGorelli) [inline pyproject.toml:187]: let's try adding to `.pre-commit-config.yaml` too?

**Comment by jorisvandenbossche** [comment]
Currently, the errors that I get are:

```
$ ruff pandas
Found 25 error(s).
pandas/core/arrays/casting.py:100:1: E501 Line too long (101 > 88 characters)
pandas/core/arrays/casting.py:111:1: E501 Line too long (99 > 88 characters)
pandas/core/arrays/casting.py:144:70: W292 No newline at end of file
pandas/core/computation/pytables.py:247:25: E713 Test for membership should be `not in`
pandas/core/generic.py:10825:52: B026 Star-arg unpacking after a keyword argument is strongly discouraged. It only works when the keyword parameter is declared after all parameters supplied by the unpacked sequence, and this change of ordering can surprise and mislead readers.
pandas/io/formats/latex.py:57:1: B024 `RowStringConverter` is an abstract base class, but it has no abstract methods
pandas/io/formats/style.py:2250:24: E714 Test for object identity should be `is not`
pandas/io/parsers/python_parser.py:1001:21: E741 Ambiguous variable name: `l`
pandas/plotting/_matplotlib/core.py:458:5: B027 `MPLPlot` is an empty method in an abstract base class, but has no abstract decorator
pandas/plotting/_matplotlib/core.py:667:5: B027 `MPLPlot` is an empty method in an abstract base class, but has no abstract decorator
pandas/plotting/_matplotlib/core.py:1858:20: E741 Ambiguous variable name: `l`
pandas/tests/arithmetic/test_datetime64.py:208:9: E741 Ambiguous variable name: `l`
pandas/tests/arithmetic/test_datetime64.py:211:13: E741 Ambiguous variable name: `l`
pandas/tests/frame/test_constructors.py:2481:22: E713 Test for membership should be `not in`
pandas/tests/indexes/common.py:35:5: F401 `pandas.core.api.UInt64Index` imported but unused
pandas/tests/indexes/period/test_indexing.py:781:20: E713 Test for membership should be `not in`
pandas/tests/indexes/period/test_indexing.py:782:20: E713 Test for membership should be `not in`
pandas/tests/io/formats/test_format.py:143:13: E741 Ambiguous variable name: `l`
pandas/tests/io/test_html.py:346:59: B026 Star-arg unpacking after a keyword argument is strongly discouraged. It only works when the keyword parameter is declared after all parameters supplied by the unpacked sequence, and this change of ordering can surprise and mislead readers.
pandas/tests/plotting/test_datetimelike.py:1024:16: E741 Ambiguous variable name: `l`
pandas/tests/plotting/test_datetimelike.py:1048:16: E741 Ambiguous variable name: `l`
pandas/tests/plotting/test_datetimelike.py:1065:16: E741 Ambiguous variable name: `l`
pandas/tests/plotting/test_datetimelike.py:1089:16: E741 Ambiguous variable name: `l`
pandas/tests/test_problem.py:8:5: F841 Local variable `res` is assigned to but never used
pandas/util/_doctools.py:80:21: E741 Ambiguous variable name: `l`
```

Are that all things we are OK with to fix, or do we want to ignore those rules instead?

(it seems we are currenlty not ignoring E741 (ambiguous short variable name) and E713 (Test for membership should be `not in`), but so ruff seems to find a few more of those compared to flake8?)

**Comment by MarcoGorelli** [comment]
I think it'd be good to fix them - is it OK if I add a commit to do that? I'd be pretty keen to get this in and remove flake8, it'd be a nice improvement to dev workflow

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me, big +1 from me (though I added a commit to fixup the remaining issues, so probably best if someone else takes a look too)

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -66,16 +70,6 @@ repos:
...
--   repo: https://github.com/PyCQA/flake8
-    rev: 6.0.0
-    hooks:
-    -   id: flake8
-        # Need to patch os.remove rule in pandas-dev-flaker
-        exclude: ^ci/fix_wheels.py
-        additional_dependencies: &flake8_dependencies
-        - flake8==6.0.0
-        - flake8-bugbear==22.7.1
-        - pandas-dev-flaker==0.5.0
```
Comment (jorisvandenbossche) [inline .pre-commit-config.yaml:78]: Completely removing flake8 here means we also stop using pandas-dev-flaker. I am not familiar with it, but are there things we would like to keep from that? Is it possible to run flake8 _only_ with the pandas-dev-flaker checks?

File: pandas/tests/io/test_html.py [hunk 3]
```diff
@@ -406,9 +406,7 @@ def test_invalid_table_attrs(self, banklist_data):
             )
 
     def _bank_data(self, path, *args, **kwargs):
```
Comment (jorisvandenbossche) [inline pandas/tests/io/test_html.py:408]: Remove the *args here as well?

```suggestion
    def _bank_data(self, path, **kwargs):
```

File: .pre-commit-config.yaml (hunk 2 above)
Comment (MarcoGorelli) [reply .pre-commit-config.yaml:78]: I looked into that but it's not so simple...I think the simplest thing would be to revert the pandas-dev-flaker PR https://github.com/pandas-dev/pandas/pull/50519

File: pyproject.toml [hunk 4]
```diff
@@ -184,6 +184,80 @@ exclude = '''
...
+  "doc/build/*.py",
+  "doc/temp/*.py",
+  ".eggs/*.py",
+  "versioneer.py",
+  # exclude asv benchmark environments from linting
+  "env",
+]
+
+[tool.ruff.per-file-ignores]
+# ignoring multiline is not yet supported (https://github.com/charliermarsh/ruff/issues/1052)
```
Comment (jorisvandenbossche) [inline pyproject.toml:253]: This one should be fixed now (since 0.0.203)

File: pyproject.toml [hunk 5]
```diff
@@ -184,6 +184,80 @@ exclude = '''
...
+"pandas/util/__init__.py" = ["F401"]
+"pandas/tests/extension/base/__init__.py" = ["F401"]
+# undefined name with global (https://github.com/charliermarsh/ruff/issues/960)
+"pandas/io/clipboard/__init__.py" = ["F821"]
```
Comment (jorisvandenbossche) [inline pyproject.toml:257]: Should also be fixed since 0.0.171

File: .pre-commit-config.yaml (hunk 2 above)
Comment (jorisvandenbossche) [reply .pre-commit-config.yaml:78]: OK, I see, that's indeed also a good option.

File: pyproject.toml (hunk 5 above)
Comment (MarcoGorelli) [reply pyproject.toml]: yup, thanks! all the per-file-ignores ones could be removed in fact, the Literal one was just due to `Literal` having been imported from `pandas._typing` rather than `typing`, so a static tool doesn't recognise it

Comment (jorisvandenbossche) [reply pyproject.toml]: Cool, thanks!

**Comment by jorisvandenbossche** [comment]
> (though I added a commit to fixup the remaining issues, so probably best if someone else takes a look too)

All code changes look fine to me.

File: pyproject.toml [hunk 6]
```diff
@@ -184,6 +184,71 @@ exclude = '''
...
+  # pycodestyle
+  "E",
+  "W",
+  # flake8-bugbear
+  "B",
+]
+
+ignore = [
+  # space before : (needed for how black formats slicing)
+  # "E203",  # not yet implemented
```
Comment (jorisvandenbossche) [inline pyproject.toml:202]: It might seem that ruff doesn't implement full of flake8, since some of those rules we ignore are not implemented, but the context is that those were not prioritized, exactly because those are not needed when using black (https://github.com/charliermarsh/ruff/issues/1527#issuecomment-1368400284)

**Comment by MarcoGorelli** [comment]
> > (though I added a commit to fixup the remaining issues, so probably best if someone else takes a look too)
> 
> All code changes look fine to me.

Cool - good to merge, or shall we wait for a 3rd opinion?

**Comment by mroeschke** [comment]
+1 to move to a faster style checker. Probably comfortable merging https://github.com/pandas-dev/pandas/pull/50519 first so we don't lose the pandas style checker checks in the meantime

File: pyproject.toml [hunk 7]
```diff
@@ -184,6 +184,71 @@ exclude = '''
...
+
+  # Additional checks that don't pass yet
+  # `zip()` without an explicit `strict=` parameter
+  "B905",
+  # Within an except clause, raise exceptions with ...
+  "B904",
+]
+
+exclude = [
+  "*.pyi",
```
Comment (twoertwein) [inline pyproject.toml:242]: At least on 0.0.211, the ignore for pyi files is not needed.

File: pyproject.toml [hunk 8]
```diff
@@ -184,6 +184,71 @@ exclude = '''
 )
 '''
 
+[tool.ruff]
+line-length = 88
```
Comment (twoertwein) [inline pyproject.toml:188]: Could also add

```
update-check = false  # there is a release every day
target-version = "py38"  # might only be important when we add ruff's pyupgrade features
fix = true  # compared to flake8/pylint, ruff has auto-rewrites to fix many rules
```

(btw. thank you for opening this PR - that's how I learned about ruff, I like it a lot!)

Comment (MarcoGorelli) [reply pyproject.toml:188]: are we sure we want `fix=true` here? because it might rewrite things which people don't want, such as 

https://github.com/pandas-dev/pandas/blob/2da7128e3436db520b96ccb07e56811a4bfcc72f/pandas/tests/frame/indexing/test_where.py#L824

(if there wasn't the `noqa`, then `--fix` would change `==False` to `is False`, though that's not the same in this context)

I think it'd be safer to let devs fix the issues? if you already have the other linting tools, then there's shouldn't be too many to deal with

Comment (twoertwein) [reply pyproject.toml:188]: Given that ruff is still adding rules and removing false positives, it can make sense not to let it autofix. If it doesn't report issues (which should be the case on the CI), it also won't rewrite anything. Just thought that similar to black, it might be nice to also fix things :) I have no strong opinion on adding/not adding it.

File: environment.yml [hunk 9]
```diff
@@ -78,12 +78,11 @@ dependencies:
   - black=22.10.0
   - cpplint
   - flake8=6.0.0
-  - flake8-bugbear=22.7.1 # used by flake8, find likely bugs
   - isort>=5.2.1  # check that imports are in the right order
   - mypy=0.991
   - pre-commit>=2.15.0
-  - pycodestyle  # used by flake8
   - pyupgrade
+  - ruff=0.0.205
```
Comment (twoertwein) [inline environment.yml:85]: Needs to be synced with pre-commit (ideally enforced or comments in both files)

File: pyproject.toml [hunk 10]
```diff
@@ -184,6 +184,78 @@ exclude = '''
...
+  # Use of functools.lru_cache or functools.cache on methods can lead to memory leaks.
+  "B020",
+  # Loop control variable overrides iterable it iterates
+  "B023",
+  # Functions defined inside a loop must not use variables redefined in the loop
+  # "B301",  # not yet implemented
+
+  # Additional checks that don't pass yet
+  # `zip()` without an explicit `strict=` parameter
+  "B905",
```
Comment (twoertwein) [inline pyproject.toml:244]: That can probably be removed since we now set the python version

File: requirements-dev.txt [hunk 11]
```diff
@@ -55,12 +55,11 @@ asv>=0.5.1
 black==22.10.0
 cpplint
 flake8==6.0.0
-flake8-bugbear==22.7.1
 isort>=5.2.1
 mypy==0.991
 pre-commit>=2.15.0
-pycodestyle
 pyupgrade
+ruff==0.0.205
```
Comment (twoertwein) [inline requirements-dev.txt:62]: Need to sync with pre-commit

File: environment.yml (hunk 9 above)
Comment (twoertwein) [reply environment.yml:85]: Should ruff actually be here? People are supposed to run pre-commit.

**Comment by MarcoGorelli** [comment]
good points @twoertwein , thanks - we OK to get this in now?

There's a few more checks that can be enabled (e.g. `TID252` allows to get rid of `absolufy-imports`), but perhaps let's keep that to a follow-up?

Regarding removing linters from the environment file - yeah I'd be totally in favour of that, let's just touch base on this in the call and then do it

**Comment by MarcoGorelli** [comment]
> Regarding removing linters from the environment file - yeah I'd be totally in favour of that, let's just touch base on this in the call and then do it

Just got:

```
  error    libmamba Could not solve for environment specs
      Encountered problems while solving:
        - nothing provides requested ruff 0.0.217**
      
      The environment can't be solved, aborting the operation
      
  critical libmamba Could not solve for environment specs
  Error: Failed to execute ["bash","-c","micromamba create -n test --strict-channel-priority -y --log-level warning -f /home/runner/work/pandas/pandas/environment.yml"]
```

Let's just spare ourselves with extra headaches and just not include it in the environment.yml file?

File: pyproject.toml (hunk 8 above)
Comment (jorisvandenbossche) [reply pyproject.toml:188]: In theory black only reformats (the ast stays the same), and never actually rewrites. So that's less intrusive to let it autofix. 

I would also first get some experience with using ruff, we could always consider later to add autofixing.

File: environment.yml (hunk 9 above)
Comment (jorisvandenbossche) [reply environment.yml]: > Should ruff actually be here? People are supposed to run pre-commit.

I think we don't require people to use pre-commit (only recommend it in our contributing guide). In general the environment seems to contain all tools, it also includes black, isort, etc

**Comment by jorisvandenbossche** [comment]
conda-forge is at 0.0.215 (https://anaconda.org/conda-forge/ruff), so you can use that version

**Comment by jorisvandenbossche** [comment]
> Regarding removing linters from the environment file - yeah I'd be totally in favour of that, let's just touch base on this in the call and then do it

If you want to discuss that, can you open an issue about it? (that can be done in advance of the call)

**Comment by MarcoGorelli** [comment]
sure, have reverted to 0.0.215

there's https://github.com/pandas-dev/pandas/issues/46561 from sometime ago, just never got around to discussing it in calls, perhaps worth revisiting if there's time

**Review by twoertwein — APPROVED** [review]

**Comment by MarcoGorelli** [comment]
Let's get this in - thanks @jorisvandenbossche !

**Comment by jorisvandenbossche** [comment]
And thanks you for much of the follow-up work!

File: pyproject.toml [hunk 12]
```diff
@@ -184,6 +184,76 @@ exclude = '''
...
+  # tests use comparisons but not their returned value
+  "B015",
+  # false positives
+  "B019",
+  # Loop control variable overrides iterable it iterates
+  "B020",
+  # Function definition does not bind loop variable
+  "B023",
+  # Functions defined inside a loop must not use variables redefined in the loop
+  # "B301",  # not yet implemented
```
Comment (charliermarsh) [inline pyproject.toml:240]: Sorry for the drive-by, but I believe this was removed from `flake8-bugbear` when it went Python 3-only: https://github.com/PyCQA/flake8-bugbear/pull/182

File: pyproject.toml (hunk 5 above)
Comment (charliermarsh) [reply pyproject.toml]: Looking at `pandas/core/reshape/merge.py`, this may no longer be relevant, but as a heads up, we do now support a setting specifically for this use-case, to mark modules as "`typing` redefinitions`.

Issue that motivated it (from Airflow): https://github.com/charliermarsh/ruff/issues/1744
Docs: https://github.com/charliermarsh/ruff#typing-modules

**Comment by charliermarsh** [comment]
This is so cool! Thank you for trying Ruff. I had no idea this was happening. You all should've been bugging me way more for fixes :)

(And sorry again for the PR spam. Just never thought I'd see Ruff in a project like Pandas :))

File: pyproject.toml (hunk 12 above)
Comment (MarcoGorelli) [reply pyproject.toml:240]: cool, thanks for your input! yes there's a few things that could be addressed in a follow-up, such as this one

**Comment by jorisvandenbossche** [comment]
> Just never thought I'd see Ruff in a project like Pandas :)

And thanks a lot to _you_ for Ruff, it's a great project making the python linting experience so much better!
